### PR TITLE
Upgrade Prometheus, Alertmanager and Thanos

### PR DESCRIPTION
**Pull request by pracucci**
**What this PR does**:
To unblock https://github.com/cortexproject/cortex/pull/4085 we need to upgrade `prometheus/common` and Alertmanager which, due to dependencies, require to upgrade Prometheus, Thanos and some other dependencies. This is the reason why in this PR you will see a bunch of dependencies updated.

I've checked the changes in all core dependencies (AWS, GCP, Alertmanager, Prometheus, Thanos) and everything looks safe to me.

**Which issue(s) this PR fixes**:
N/A

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: go.mod
Comment (pracucci) [inline go.mod:88]: My plan is to keep this, so that we'll not block this PR by Thanos. I will remove this replace once Thanos PR is merged.

**Review by pstibrany — APPROVED** [review]
Cortex changes LGTM.

**Comment by pracucci** [comment]
I've converted into a draft because, while reviewing https://github.com/thanos-io/thanos/pull/3937, I've spot some issues which we need to fix first.

**Comment by pracucci** [comment]
> I've converted into a draft because, while reviewing [thanos-io/thanos#3937](https://github.com/thanos-io/thanos/pull/3937), I've spot some issues which we need to fix first.

Opened a Thanos PR https://github.com/thanos-io/thanos/pull/4095

**Comment by pracucci** [comment]
> > I've converted into a draft because, while reviewing [thanos-io/thanos#3937](https://github.com/thanos-io/thanos/pull/3937), I've spot some issues which we need to fix first.
> 
> Opened a Thanos PR [thanos-io/thanos#4095](https://github.com/thanos-io/thanos/pull/4095)

Merged and updated vendored Thanos in this PR. The risky change here is https://github.com/thanos-io/thanos/pull/3937 but I've reviewed and, despite complex, I haven't noticed any issue (except for the race condition in the `pool.BuckedBytes`, which has been merged.

**Pull request merged**
